pyHalo/realization_extensions.py

```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
+            
+            # get the realization at redshift z
+            realization_at_redshift, indexes = realization_at_z(realization, zi, center_x, center_y, max_rendering_range)
+            
+            # get an instance of LensModel at redshift z
+            args = realization_at_redshift.lensing_quantities(add_mass_sheet_correction=False)
+            lens_model_list_at_plane, kwargs_lens_at_plane = args[0], args[2]
+            
+            delta_zi = 0.02
+            #delta_zi= lens_plane_redshifts[1]-lens_plane_redshifts[0]   
+            V = cosmo_geometry.volume_element_comoving(zi, delta_z=delta_zi, radius=max_rendering_range)
+          
+            f_halos =  LMF.mass_fraction_in_halos(zi, mlow, mhigh) # this is where mlow and mhigh are necessary
+            
+            # M is black hole mass (solar masses)
+            
+            _nsmooth = f_halos*rho_DM*(V/M)
+            _nclumpy = (1-f_halos)*rho_DM*(V/M)
+            
+            Nclumpy = int(np.random.poisson(_nclumpy))
+            Nsmooth = int(np.random.poisson(_nsmooth))
+            
+    
+            if Nsmooth > 0:
+                coord_x_smooth, coord_y_smooth,  z_shifts_smooth, masses_smooth = self._smooth_pbh (max_rendering_range, Nsmooth, center_x, center_y, zi, M)
+                pbh_x_coordinates = np.append(pbh_x_coordinates, coord_x_smooth)
+                pbh_y_coordinates = np.append(pbh_y_coordinates, coord_y_smooth)
+                pbh_redshifts += z_shifts_smooth
+                pbh_masses = np.append(pbh_masses, masses_smooth)
+            
+            # we can only generate PBH around halos if there are halos, hence the second condition
+            if Nclumpy > 0 and len(lens_model_list_at_plane): 
+                coord_x_clumpy, coord_y_clumpy, z_shifts_clumpy, masses_clumpy = self._clumpy_pbh (lens_model_list_at_plane, center_x, center_y, 
+                     kwargs_lens_at_plane, Nclumpy, max_rendering_range, npix, zi, M)
+                pbh_x_coordinates = np.append(pbh_x_coordinates, coord_x_clumpy)
+                pbh_y_coordinates = np.append(pbh_y_coordinates, coord_y_clumpy)
+                pbh_redshifts += z_shifts_clumpy
+                pbh_masses = np.append(pbh_masses, masses_clumpy)
+                
+        return pbh_x_coordinates, pbh_y_coordinates, pbh_redshifts, pbh_masses
+    
+    def pbh_around_image(self, compfrac, M, image_index, realization, cosmology, lens_cosmo, cosmo_geometry, 
+                     zlens, zsource, x_interp_list, y_interp_list):
+    
+    
+        lens_plane_redshifts = self._realization.unique_redshifts
+        
+        max_rendering_range_base = 0.25 # arcsec
+        npix = 300
+        
+        # defining things for later
+        conangle = cosmo_geometry.cone_opening_angle #arcsec 
+        LMF = LensingMassFunction(cosmology, zlens, zsource, cone_opening_angle=conangle)
+        mlow = 1e6
```

Comment:
there are keyword arguments stored in the realization instance (self._realization) that define the mass range, access them with profile_args = self._realization._prof_params and then mlow = 10**profile_args['log_mlow'], etc.